1382 split asset page into several views

Muhammad-Moiz626 opened the pull request:
## Description

Split the Asset page into Graph, Properties, AuditLog and Status.

## Change Log

- Created Structure for Assets using Vega Library.
- Made Asset Context page as default page.
- Added Map to the Context Page.
- Separated the logic for show all sensors and edit flex context.
- Removed the extra stuff from the Asset main page.
- Created separate pages for Properties, Logs, Status and Audit Log.
- Added BreadCrumb for the Asset pages.
- Updated some tests as well based on the new routing.

## How to test

- Login to flexmeasures.
- Go to Assets listing page.
- Click on one of the assets from the list.
- Make sure the page opened is not full of a lot of information.
- Make sure there is an additional dropdown in Breadcrumb that says `Views`.

## Related Items

Closes #1382 
This PR is dependent on the merging of [PR #1369](https://github.com/FlexMeasures/flexmeasures/pull/1369)

---

- [x] I agree to contribute to the project under Apache 2 License. 
- [x] To the best of my knowledge, the proposed patch is not based on code under GPL or other license that is incompatible with FlexMeasures

nhoening reviewed (changes requested):
Cool!

Now that you have merged the context PR into this one, do you want to continue only with this?

For the breadcrumbs:
-  I would like it to not say "Views" always, but display the current view name, e.g. "Overview", "Graphs", "Properties", etc.
- The context view seems not to be reachable now, should be one of the views.

Let's get rid of the old page. The context page should become the new default asset URL (available /assets/ID).

The status page has a "Back to asset page" button, which can go away now.

The properties page has a button to create and to delete. Create would not be needed (is on context page now), and delete should actually also rather be there, next to the existing actions.

Flix6x commented:
Will selecting a different view reload the entire page or just update the contents visible on the asset page?

nhoening commented:
This reloads the page. Your idea could be done at a later time, but I deem it not so critical that we must Ajax-ify all of this now.

nhoening commented on flexmeasures/ui/templates/crud/asset_graph.html:
```diff
@@ -0,0 +1,1208 @@
+        renderFlexContextForm(); // Initial render of flex context form
+    });
+
+    flexContexFormModal.addEventListener('hidden.bs.modal', function () {
```
This would not be needed here anymore, as that modal dialogue is now on the context page (I believe, at least not here).

I guess one part of remaining work is to clean each view from stuff it doesn't need.

Same goes for `renderFlexContextForm()` above and `handleFlexSelectChange()` below.

Muhammad-Moiz626 replied:
In this case I think we should wait for Joshua to finish his work first and then I'll merge my code on top of that.

nhoening replied:
Probably for the best. He could even do some of the cleaning, as he knows those parts better. Or in collaboration.

nhoening reviewed (changes requested):
One more comment, about code this time. 
- I believe we can clean out code in the views, see inline comment. Next to the old asset page not being needed anymore.
- Also, we are still doing some finalizing on context editing in #1384 and #1390, sorry about that. I suggest we merge those soon, and then the differences can be worked out. You could also ask Joshua to sort out what is the latest state, then.

nhoening commented:
By the way, tests are failing, maybe just simple UI tests, but it would be good to fix them here, as well.

Flix6x commented:
> This reloads the page.

Should we redirect the old page to the new view showing the graphs? As a way to avoid dead links, e.g. in plugins.

nhoening commented:
I think we should use the context as new main page.

It will lighten the load on servers as often, users are not interested in graphs, but we still load data.

I don't believe too many plugins will be affected, and it's easy to adapt.

Flix6x commented:
I figured you'd already update all links within FM to point to the context view.

But as long as there is a redirect to avoid dead links, that's my main concern.

nhoening commented on flexmeasures/ui/utils/breadcrumb_utils.py:
```diff
@@ -6,12 +6,71 @@
+                },
+                {
+                    "url": url_for("AssetCrudUI:graph", id=entity.id),
+                    "name": "Graph",
```
I'd call this "Graphs"

nhoening commented on flexmeasures/ui/templates/crud/asset_properties.html:
```diff
@@ -0,0 +1,335 @@
+                                <div class="form-group">
+                                    <label class="control-label">Location</label>
+                                    <small>(Click map to edit latitude and longitude in form)</small>
+                                    <div id="mapid"></div>
```
This stays empty, maybe some JavaScript is not working anymore. It might have been broken on main already, though, if that is the case we could open a new issue for this.

Muhammad-Moiz626 replied:
I have merged the main and fixed the issue for map as well.

nhoening commented on flexmeasures/ui/templates/crud/asset_properties.html:
```diff
@@ -0,0 +1,335 @@
+                                </div>
+
+                                <div class="form-group">
+                                    <label for="flex_context" class="col-sm-6 control-label">Flex Context: </label>
```
The flex context editing option should move to the context page.

nhoening commented on flexmeasures/ui/templates/crud/asset_properties.html:
```diff
@@ -0,0 +1,335 @@
+                                    </button>
+                                </div>
+
+                                <div class="form-group">
```
The graph editing option already lives on the graph page, so this can go.

nhoening commented on flexmeasures/ui/templates/crud/asset_properties.html:
```diff
@@ -0,0 +1,335 @@
+        </div>
+   <div class="col-md-8">
+     
+
```
Actually, here it would be great to show the properties that are in the edit form on the left, just in a normal info table, above the other tables for sensors and child assets.

Should be easy as all this information is already loaded for the edit form.

nhoening reviewed (changes requested):
One thing to look into is this error when I click on "Add Child":

![Screenshot from 2025-04-04 16-57-43](https://github.com/user-attachments/assets/16225a46-13b5-4d72-b678-94c5f13b6802)

The map is still not showing tiles, while that works fine on the dashboard, that must be a bug...

In better news, the styling suddenly looks better for me now, hope it stays this way :)

nhoening reviewed (changes requested):
Thanks, I believe we now need to click through everything to make sure it's still working. I found a few problems on the context page:

- I see an error in the console when loading: "Map container not found.", then later: "Map container is already initialized". The feature itself is working, but this seems weird.
- Editing the flex context is not working - no fields to choose from. The console has this error: "JSON.parse: unexpected end of data at line 1 column 1 of the JSON data", happening in the code below. Maybe ask Joshua if he can spot what went wrong while porting, not obvious to me.

```
    let availableUnitsRawJSON = "";
    availableUnitsRawJSON = availableUnitsRawJSON.replace(/'/g, '"');
    const availableUnits = JSON.parse(availableUnitsRawJSON);
```


On the graph page:
- When I click the button to edit graph titles, the console says: "ReferenceError: selectFlexField is not defined"

On properties page:
Editing seems to work, but saving brings me to the context page (well, the default asset page), but it would be best to stay on the properties page. Just a fix to the redirect or something similar, I guess.


And I suggest one larger improvement on the context page, now that I have seen it with assets that don't have a rich hierarchy - I would like to show the map not in a modal, but always. So in this page, we show two major things (structure and location) - which should be togglable by tabs (https://getbootstrap.com/docs/5.0/components/navs-tabs/#tabs).
And I'd show the location tab per default if there is no parents or children, and preferrably the structure tab otherwise.

nhoening commented on flexmeasures/ui/templates/crud/asset_graph.html:
```diff
@@ -0,0 +1,746 @@
+            spinnerElement.classList.add('hidden-important');
+            senSearchResEle.style.display = 'block';
+        } catch (error) {
+            showToast("Failed to filter/search sesnors", "error");
```
Huh, there is a pre-existing typo: "sesnors". Can you fix it? It appears twice, I'll also comment on the other one.

Also, we can just say "search" here.

nhoening commented on flexmeasures/ui/templates/crud/asset_graph.html:
```diff
@@ -0,0 +1,746 @@
+            spinnerElement.classList.add('hidden-important');
+            apiSensorsListElement.style.display = 'block';
+        } catch (error) {
+            showToast("Failed to filter/search sesnors", "error");
```
other typo here. Also we can just say "filter" here

nhoening reviewed (changes requested):
Very close, thanks! I only found a few things.

The map overlays Toast messages, see Screenshot below. Not sure where, but I guess the z-index needs adaptation.

![Screenshot from 2025-04-11 16-16-22](https://github.com/user-attachments/assets/5e0895c8-a7fb-4525-bb7e-1cc612697669)

On the context page, I can change the marker position, but that should not be possible. 

On the properties page, the first table should not be a table with the asset as only row.
Best to do a simple striped table (like e.g. on the sensor page) with each main field on the left, and on the right a table just for the attributes (each attribute one row).

Muhammad-Moiz626 commented:
```
On the properties page, the first table should not be a table with the asset as only row.
Best to do a simple striped table (like e.g. on the sensor page) with each main field on the left, and on the right a table just for the attributes (each attribute one row).
```

I don't get this. On the sensor page there are defined attributes but here there are multiple things in the attributes. Even logs are also there. How do I show them?

nhoening commented:
> ```
> On the properties page, the first table should not be a table with the asset as only row.
> Best to do a simple striped table (like e.g. on the sensor page) with each main field on the left, and on the right a table just for the attributes (each attribute one row).
> ```
> 
> I don't get this. On the sensor page there are defined attributes but here there are multiple things in the attributes. Even logs are also there. How do I show them?

Yes, I did not consider that in the simulations we use attributes to store a lot of data.
For the table on the right (for attributes), we should thus take inspiration from our simulation code, where we show attributes in a table (e.g. https://simulations.seita.energy/flexmeasures-smartbuildings/scenarios/1401). We can choose to add a vertical scroll bar.

Does that help? I believe you were only asking about the table for attributes on the right.

Muhammad-Moiz626 commented:
I have updated the tables as per the suggestion.

P.S. I have fixed the `spinner` issue as well. Somehow on my local it shows a spinner at the start and hides soon after the graph is shown. But I'm not sure if this is the correct behaviour we want. Please verify that on your end.

nhoening commented on flexmeasures/ui/templates/crud/asset_properties.html:
```diff
@@ -0,0 +1,379 @@
+                                        <td>{{ asset.latitude }}</td>
+                                        <td>{{ asset.longitude }}</td>
+                                        <td>{{ asset.parent_asset.name }}</td>
+                                        <td>{{ asset.parent_asset.generic_asset_type.name }}</td>
```
Here I encountered:

```
  File "/home/nicolas/workspace/seita/flexmeasures/flexmeasures/ui/templates/crud/asset_properties.html", line 154, in block 'divs'
    <td>{{ asset.parent_asset.generic_asset_type.name }}</td>
  File "/home/nicolas/envs/test-fm-pyproject2.toml/lib/python3.10/site-packages/jinja2/environment.py", line 490, in getattr
    return getattr(obj, attribute)
jinja2.exceptions.UndefinedError: 'None' has no attribute 'generic_asset_type'
```

I believe that asset did not have a parent.

nhoening commented on flexmeasures/ui/crud/assets/views.py:
```diff
@@ -312,13 +308,68 @@ def delete_with_data(self, id: str):
+        )
+
+    @login_required
+    @route("/<id>/graph")
```
I'd rather name this `/<id>/graphs`, with an "s"

nhoening commented on flexmeasures/ui/templates/crud/asset_properties.html:
```diff
@@ -0,0 +1,379 @@
+                <div class="row">
+                    <div class="col-md-7">
+                        <div class="table-responsive">
+                            <table class="table table-striped nav-on-click w-100 mx-auto" title="View Asset Properties">
```
This table should not have the fancy smart table capabilities like "Filter records", sorting, pagination

Make it a simple table, and also vertical, much like the one on the right for attributes. With two columns, "Property" and "Value".

(Maybe `render_attributes()` can be used for that or copied?)

nhoening reviewed (changes requested):
Thanks. A few small issues / ideas.

Some tests (test_asset, test_asset_crud) are failing, hopefully easy to fix as they test content usually, and you basically only moved content between pages.

nhoening commented on flexmeasures/ui/crud/assets/views.py:
```diff
@@ -362,9 +362,24 @@ def properties(self, id: str):
+            "Name": asset.name,
+            "Latitude": asset.latitude,
+            "Longitude": asset.longitude,
+            "Parent Asset": (
```
I would like to summarize parent asset info in one row, maybe even clickable, like  `<a href="/assets/ParentID">Asset name (type)</a>`

nhoening reviewed (changes requested):
Awesome, this looks finished, just one small wish.

And of course the tests.

nhoening reviewed (changes requested):
Everything works nicely!

Before merging, can we keep [the dates-validation test](
https://github.com/FlexMeasures/flexmeasures/pull/1387/files/042493e6c3f2f95b1bad6ee33db1a45e1a889363..6f82c456ec7aa575e8bbf8a68f1f38b3099896a9#diff-e8f35c7ffbe951f6fe917fe0b2c3e1cba114d9b64e11ca8477199cbaa6cdf135L69), by simply linking it to the graphs page? I believe the functionality is still there, and only moved to a different page.
